HPOS: Add background sync

coreymckrill opened the pull request:
### Changes proposed in this Pull Request:

Adds a "Background Sync" feature that can run in parallel with, or in place of the "real-time" order data sync that is enabled by the "Compatibility Mode" checkbox on the WC Features screen. For now, the background sync mode has no UI of its own, it is either activated by simply having the normal data sync mode also enabled, or if that is disabled, it can be activated separately using a filter hook.

Background Sync can either be run at a set interval, or in "continuous" mode, where the data synchronizer is always enqueued in the batch processor so that it will always catch changes to order data as they come in (but still process them asynchronously so that order operations remain performant).

Also changes the "Sync x pending orders" message to a button that, when clicked, will immediately start a syncing process for those orders.

Fixes #39626

### How to test the changes in this Pull Request:

There are several different modes to test in this PR, but the action you perform for each test is essentially the same. As described in the reproduction steps in #39626, you just delete an order from the wc_orders table using a raw SQL query so that you bypass the hooks and filters that would normally run when you delete an order through the WC API. WooCommerce will then detect that the posts and orders tables are out of sync. What it does after that is what we're testing here.

#### Real-time sync on, background sync on

1. After checking this branch out, go to _WP Admin > WooCommerce > Settings > Advanced > Features_, and make sure that **Order data storage** is set to "legacy" (i.e. posts table) and the compatibility mode box is checked (i.e. data sync is _on_). In a separate tab, open up _WP Admin > Tools > Scheduled Actions_ and run any pending actions named `wc_schedule_pending_batch_processes` so there aren't any left in the queue.
2. As in #39626, manually delete an order and then refresh the Features screen. You should see the message that you can't switch authoritative tables until the data is in sync, and a button saying "Sync 1 order".
3. Switch to the Scheduled Actions tab and refresh. You should see a scheduled event named `woocommerce_custom_orders_table_background_sync` that will run in about an hour. Click the row action to run the event now. That should spawn a new `wc_schedule_pending_batch_processes` event, which may run immediately, or may be delayed by up to a minute (this is how Action Scheduler works, not part of this PR)
4. Back on the Features screen, you should now see a "Currently syncing" message, or nothing if the sync has already completed.
5. The orders should be back in sync now, so the one you deleted should be restored at this point.
6. Delete the order again and refresh the Features screen. This time, click the "Sync 1 order" button. The message should change to "Currently syncing". On the Scheduled Actions screen, you should see the `wc_schedule_pending_batch_processes` event queued up to run immediately.
7. Once the event has finished, the Features screen should show that everything is back in sync.

#### Real-time sync **off**, background sync on, "interval" mode

Ok, so now we're going to test background sync by itself, while the regular sync is disabled. Background sync has two active modes, "interval" and "continuous". "interval" is the default, so we'll test that first.

1. On the Features screen, uncheck the compatibility mode checkbox and save your changes. On the Scheduled Actions screen, the `woocommerce_custom_orders_table_background_sync` event should now be gone. All syncing is now disabled.
2. To re-enable just background sync, run the following WP CLI command:
    ```bash
   wp option update woocommerce_custom_orders_table_background_sync_mode interval
    ```
3. Now on the Features screen, it should say "Background sync enabled" and on the Scheduled Actions screen, the `woocommerce_custom_orders_table_background_sync` event should be there again.
4. Run through steps 2-7 from the previous section again to test syncing with the scheduled event and syncing with the button. You should get the same results.

#### Real-time sync **off**, background sync on, "continuous" mode

1. Now run this WP CLI command:
    ```bash
    wp option update woocommerce_custom_orders_table_background_sync_mode continuous
    ```
2. On the Scheduled Actions screen, there should be a `wc_schedule_pending_batch_processes` event queued up, even though the orders are currently in sync.
3. Delete the order again, and immediately refresh the Features screen. Rather than showing the "Sync 1 order" button, it should have skipped directly to "Currently syncing".
4. Even after the event has completed and everything is back in sync, there should be another `wc_schedule_pending_batch_processes` event queued up.

#### Turn background sync back off

1. Run this WP CLI command:
    ```bash
    wp option update woocommerce_custom_orders_table_background_sync_mode off
    ```
    or this one:
    ```bash
    wp option delete woocommerce_custom_orders_table_background_sync_mode
    ```
2. There should not be a `woocommerce_custom_orders_table_background_sync` event in Scheduled Actions, nor a `wc_schedule_pending_batch_processes` event.
3. The "Background sync enabled" message should be gone.
4. If you delete the order again, you should still see a "Sync 1 order" button that you can click to manually sync the order.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message

Add a "background" sync that can run independently of the normal "real-time" HPOS data sync. Also add a button on the Features screen to trigger an order sync manually.

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

coreymckrill committed:
HPOS: Add a recurring event to check for unsynced orders

Adds an event that will get scheduled when HPOS data sync gets enabled.
The event will run every 6 hours and check for unsynced orders. If it
finds any, it will queue up the  data syncing batch processor.

Fixes #39626

coreymckrill committed:
Add button to sync orders immediately

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

coreymckrill commented:
This changeset is evidently causing many occurrences of the following error in the unit tests:

```
as_unschedule_all_actions() was called before the Action Scheduler data store was initialized
```

Looking into it...

coreymckrill committed:
Wrap AS function calls to avoid errors in unit tests

coreymckrill committed:
Use WC_Queue methods instead of AS functions

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -347,8 +355,88 @@ private function handle_data_sync_option_changed( string $feature_id ) {
+		if ( ! $has_scheduled_action ) {
+			WC()->queue()->schedule_recurring(
+				time() + HOUR_IN_SECONDS,
+				6 * HOUR_IN_SECONDS,
```
what do you think about adding the duration as a filter? so that folks who want it at a more regular frequency can do so.

coreymckrill replied:
Good idea: b488e9d480

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -347,8 +355,88 @@ private function handle_data_sync_option_changed( string $feature_id ) {
+	 * @return void
+	 */
+	private function check_for_pending_syncs() {
+		if ( ! $this->data_synchronizer->data_sync_is_enabled() ) {
```
a filter here would be nice too, which allows doing the sync in the background but having the sync otherwise disabled. this will allow use cases for folks that want to keep post and post meta tables updated, but don't want to have the sync enabled all the time to have performance

coreymckrill replied:
🤔 But if they have data sync disabled, the event to run this callback will be unscheduled, so it wouldn't run anyway. I think you're on to something that it would be interesting to allow for an... async sync option so there isn't a real-time performance hit. But I think it would take a bit more to implement, and seems like would be better addressed in a separate PR.

vedanshujain replied:
> async sync option so there isn't a real-time performance hit.

seems like we need this to support high-volume sites (such as wccom) as an intermediatory steps in turning off the sync. knowing this, would you change anything in this PR? Perhaps we can use the similar mechanism for both these things? wdyt?

coreymckrill replied:
If we go this route, it seems like we'll need to evolve the data sync setting away from a simple checkbox to something else. Or, maybe it's still a checkbox, but when it's checked, more options are revealed and you then have to decide between "real-time" syncing and periodic sync?

If we can figure out the UI for this, I think this PR gets us most of the way to the implementation. Any other thoughts you have about the UI?

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -347,8 +355,88 @@ private function handle_data_sync_option_changed( string $feature_id ) {
+			return;
+		}
+
+		$sync_status = $this->data_synchronizer->get_sync_status();
```
instead of getting_sync_status, get pending sync count would be more accurate. sync status can be outdated if there are new orders after the last time status was updated (for example, between multiple sync action runs)

coreymckrill replied:
Ah thanks, I didn't realize it was a cached value. Fixed in 9c96a84438

vedanshujain reviewed (commented):
Looks pretty neat, addded some comments. What do you think about changing the button to a link to action instead, currently button links bit out of place.

coreymckrill committed:
Add a filter to customize the sync check time interval

coreymckrill committed:
Ensure retrieved pending sync count is not a cached value

coreymckrill committed:
Change sync button style to link

coreymckrill commented:
@vedanshujain 

> What do you think about changing the button to a link to action instead, currently button links bit out of place.

Changed the button style to a link in 93375ad065. Re: "link to action instead", do you mean going to the Scheduled Actions screen? I'm concerned that this would be confusing to a lot of folks, as they wouldn't necessarily know what to do on that screen once they got there. I think it's better UX to just let them trigger the sync right there and see the message change to "Currently syncing orders..."

coreymckrill committed:
Formatting fixes

coreymckrill committed:
Fix unit test

vedanshujain commented:
Left a pending question but otherwise looks good to me (and hoping to do some more testing). Note that it allows to trigger the sync irrespective of whether the checkbox for "keeping tables in sync" is enabled or not. I think it's good, but wanted to confirm if that intentional (but maybe lets keep it even if it was unintentional).

vedanshujain commented:
@coreymckrill  in light of exploring this solution to have a permanent background sync, I have the following suggestions that I propose we apply to this PR:

We have this class called `BatchProcessingController`, which already checks for pending orders and schedules new actions as long as there is at least one pending batch returned by the DataSynchronizer class. Since it already maintains a background checker process, similar to `SYNC_CHECK_EVENT_HOOK` that we  are adding in this PR, we can

1. Add the filter to customize the time that it [schedules the normal sync action](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php#L239).
2. Add a param when scheduling 'WATCHDOG_ACTION_NAME` so that it first schedules itself for one hour, and then six hours before dequeuing. If we do that, we won't need `SYNC_CHECK_EVENT_HOOK` anymore (which is good because less code).
3. We also add a filter to the `still_pending` variable, this will allow folks to set it so that it always returns `1` and thereby always checking for new orders every minute. The time can be configured by the filter added in point 1 if someone so desires.

With these changes in place, we won't need the `SYNC_CHECK_EVENT_HOOK` anymore. Additionally, if someone wants to have a permanent background sync then they will need to add a filter to `still_pending` that we added in point 3 to always return true. Then:

```php
// To start the background data sync,  apply the filter suggested above, then
$controller = wc_get_container()->get( \Automattic\WooCommerce\Internal\BatchProcessing\BatchProcessingController::class );

$controller->enqueue_processor( \Automattic\WooCommerce\Internal\DataStores\Orders\DataSynchronizer::class );
```

To stop the background process, either remove the filter or:
```php
$controller = wc_get_container()->get( \Automattic\WooCommerce\Internal\BatchProcessing\BatchProcessingController::class );
$controller->remove_processor( \Automattic\WooCommerce\Internal\DataStores\Orders\DataSynchronizer::class );
```

To get the status (whether the processor is still enqueued):
```
$controller = wc_get_container()->get( \Automattic\WooCommerce\Internal\BatchProcessing\BatchProcessingController::class );
$controller->get_enqueued_processors(); // Check if DataSynchronizer is in array.
```

wdyt?

coreymckrill committed:
Add BatchProcessingController dependency to DataSynchronizer

coreymckrill committed:
Add background sync functionality to DataSynchronizer

coreymckrill committed:
Tweaks to watchdog scheduling in batch processor

* Adds a filter to modify the amount of time the watchdog schedule
  gets delayed when calling the scheduler using `$with_delay`
* Adds a check to ensure the scheduler doesn't overwrite an existing
  scheduled event. This is because the scheduler was getting called
  multiple times in a request, first without `$with_delay` (so the event
  would run right away) and then again with `$with_delay`, so that the
  event would then be delayed by an hour. The result was that the event
  was always scheduled for an hour later, even when we want it to run
  right away.

coreymckrill committed:
Improve query arg for running sync via UI

coreymckrill committed:
Remove sync check from COTController, add message about background sync

coreymckrill committed:
Abstract the removal of background sync events

coreymckrill committed:
Remove accidental test code

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

coreymckrill committed:
Ensure bg sync event is not scheduled when in continuous mode

coreymckrill commented:
@vedanshujain this is now refactored to take some of your suggestions above into account. I rewrote the PR description and testing instructions to reflect this. Let me know what you think!

vedanshujain commented on plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php:
```diff
@@ -99,14 +99,25 @@ public function enqueue_processor( string $processor_class_name ): void {
+		if ( $with_delay ) {
+			/**
+			 * Modify the delay interval for the batch processor's watchdog events.
+			 *
```
missing a `since` version comment.

coreymckrill replied:
Fixed in 9ce05d185345cdaa43541bc68e4f90baaa770ed8

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -171,14 +185,123 @@ public function delete_database_tables() {
+	 * @return void
+	 */
+	private function schedule_background_sync() {
+		$has_scheduled_action = WC()->queue()->search( array(
```
This an expensive call to do on every init, we need to either cache it or link it to a less frequent function. Probably both. wdyt about linking to admin_init instead of init, and then storing the epoch time in an option when this was last checked, and checking only if last_checked was more than 1 hour ago?

coreymckrill replied:
I think it's important for`handle_background_sync` to run on `init`. If it only runs on admin requests, it could be a long time between runs of that callback, and in the mean time a lot of changes to orders could happen on the front end.

I added transient caching in e63ea3bf71fc32e3f40bad4da103dbdbdd17e273, let me know what you think of that.

coreymckrill replied:
Noting that there is an [open issue](https://github.com/woocommerce/action-scheduler/issues/978) in Action Scheduler to improve the performance of this particular query.

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -171,14 +185,123 @@ public function delete_database_tables() {
+		$mode = apply_filters( 'woocommerce_custom_orders_table_background_sync_mode', 'interval' );
+
+		switch ( $mode ) {
+			case 'interval':
```
both `interval` and `continous` should be defined as constants.

coreymckrill replied:
Updated in 898679d4e1b8d7a97136ca0ecf1c446266f62d06

vedanshujain reviewed (commented):
Left few minor comments, but overall approach looks good to me! Thanks for the detailed test instructions as well. Do you think if we can add some unit tests too?

coreymckrill committed:
Add missing @since comments on new filter hooks

coreymckrill committed:
Add caching to check for scheduled bg sync events

coreymckrill committed:
Use constants for bg sync modes

coreymckrill committed:
Add unit tests

coreymckrill commented:
Added unit tests in 9a4bb675efd8d4f6174122074e06f44900cc6a4c

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -171,14 +188,179 @@ public function delete_database_tables() {
+				true
+			);
+
+			set_transient(
```
why not use options instead of transient? transients can get ejected or expired. It's more of an option rather then a cache.

coreymckrill replied:
🤔 What makes you think of this as more of an option? The mechanism seems pretty explicitly like a cache to me--we're saving a queried value to the database to avoid having to run the query every time. We _want_ the cached value to expire in case there's a pathway where the value doesn't get reset programmatically. Plus, the transient system uses a persistent object cache if it's available.

I suppose that we don't necessarily need to be setting the transient at this spot since it gets done in `background_sync_is_scheduled`...

coreymckrill replied:
Made the toggle for background sync an option instead of a filter in 52dbc901af6fd6824b2358dcdf0599df1b697a27, which allows us to avoid the need for caching, because now the event only gets scheduled/unscheduled when the option value changes.

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -89,6 +101,8 @@ public function __construct() {
 		self::add_action( 'woocommerce_refund_created', array( $this, 'handle_updated_order' ), 100 );
 		self::add_action( 'woocommerce_update_order', array( $this, 'handle_updated_order' ), 100 );
 		self::add_action( 'wp_scheduled_auto_draft_delete', array( $this, 'delete_auto_draft_orders' ), 9 );
+		self::add_action( 'init', array( $this, 'handle_background_sync' ) );
```
`init` is noisy, should we switch to admin_init instead?

coreymckrill replied:
I explained my reasoning [here](https://github.com/woocommerce/woocommerce/pull/39952#discussion_r1327735839), for sites where there isn't a user consistently logging into WP Admin, I don't think the callback would fire frequently enough.

coreymckrill replied:
Switched to the `shutdown` hook in 9c51363d87167de1b4762e7f64580a339ce6449f

coreymckrill committed:
Switch from init hook to shutdown

coreymckrill committed:
Switch bg sync settings from filters to options

coreymckrill committed:
Update unit tests

coreymckrill committed:
phpcs fix

coreymckrill committed:
Tweak "Background sync is enabled" message

Only show this if it is enabled while real-time sync is disabled.
Otherwise the message could be confusing to users who just check the
"Compatibility mode" box, since background sync doesn't have any UI.

vedanshujain reviewed (approved):
Great work! I really like that any hook to `init` is removed. Marking for internal testing and analysis.
